File: src/com/rian/difficultycalculator/beatmap/hitobject/HitCircle.java
package com.rian.difficultycalculator.beatmap.hitobject;

import com.rian.difficultycalculator.math.Vector2;

/**
 * Represents a hit circle.
 */
public class HitCircle extends HitObject {
    /**
     * @param startTime The start time of this hit circle, in milliseconds.
     * @param position  The position of this hit circle relative to the play field.
     */
    public HitCircle(double startTime, Vector2 position) {
        super(startTime, position);
    }

    /**
     * @param startTime The start time of this hit circle, in milliseconds.
     * @param x         The X position of this hit circle relative to the play field.
     * @param y         The Y position of this hit circle relative to the play field.
     */
    public HitCircle(double startTime, float x, float y) {
        super(startTime, new Vector2(x, y));
    }

    /**
     * Copy constructor.
     *
     * @param source The source to copy from.
     */
    private HitCircle(HitCircle source) {
        super(source);
    }

    @Override
    public HitCircle deepClone() {
        return new HitCircle(this);
    }
}

File: src/com/rian/osu/beatmap/hitobject/HitCircle.kt
package com.rian.osu.beatmap.hitobject

import com.rian.osu.math.Vector2

/**
 * Represents a hit circle.
 */
class HitCircle(
    /**
     * The start time of this [HitCircle], in milliseconds.
     */
    startTime: Double,

    /**
     * The position of this [HitCircle] relative to the play field.
     */
    position: Vector2,

    /**
     * Whether this [HitCircle] starts a new combo.
     */
    isNewCombo: Boolean,

    /**
     * When starting a new combo, the offset of the new combo relative to the current one.
     *
     * This is generally a setting provided by a beatmap creator to choreograph interesting color patterns
     * which can only be achieved by skipping combo colors with per-[HitObject] level.
     *
     * It is exposed via [HitObject.comboIndexWithOffsets].
     */
    comboOffset: Int
) : HitObject(startTime, position, isNewCombo, comboOffset)
